lib/gamnit/simple_2d.nit

```diff
@@ -0,0 +1,340 @@
+		var h = height * mod
+		var a = [-0.5*w, -0.5*h, 0.0]
+		var b = [ 0.5*w, -0.5*h, 0.0]
+		var c = [-0.5*w,  0.5*h, 0.0]
+		var d = [ 0.5*w,  0.5*h, 0.0]
+
+		var vertices = new Array[Float]
+		for v in [c, d, a, b] do vertices.add_all v
+		return vertices
+	end
+
+	# Coordinates of this texture on the `root` texture, with `[0..1.0]`
+	private var texture_coords: Array[Float] is lazy do
+		var a = [offset_left,  offset_bottom]
+		var b = [offset_right, offset_bottom]
+		var c = [offset_left,  offset_top]
+		var d = [offset_right, offset_top]
+
+		var texture_coords = new Array[Float]
+		for v in [c, d, a, b] do texture_coords.add_all v
+		return texture_coords
+	end
+end
+
+# Graphic program to display simple models with a texture, translation, rotation and scale
+class Simple2dProgram
+	super GamnitProgramFromSource
+
+	redef var vertex_shader_source = """
+		// Vertex coordinates
+		attribute vec4 coord;
+
+		// Vertex color tint
+		attribute vec4 color;
+
+		// Vertex translation
+		attribute vec4 translation;
+
+		// Vertex scaling
+		attribute float scale;
+
+		// Vertex coordinates on textures
+		attribute vec2 tex_coord;
+
+		// Model view projection matrix
+		uniform mat4 mvp;
+
+		// Rotation matrix
+		uniform mat4 rotation;
+
+		// Output for the fragment shader
+		varying vec4 v_color;
+		varying vec2 v_coord;
+
+		void main()
+		{
+			v_color = color;
+			gl_Position = (vec4(coord.xyz * scale, 1.0) * rotation + translation) * mvp;
+			v_coord = tex_coord;
+		}
+		""" @ glsl_vertex_shader
+
+	redef var fragment_shader_source = """
+		precision mediump float;
+
+		// Does this object use a texture?
+		uniform bool use_texture;
+
+		// Texture to apply on this object
+		uniform sampler2D texture;
+
+		// Input from the vertex shader
+		varying vec4 v_color;
+		varying vec2 v_coord;
+
+		void main()
+		{
+			if(use_texture) {
+				gl_FragColor = v_color * texture2D(texture, v_coord);
+				if (gl_FragColor.a == 0.0) discard;
+			} else {
+				gl_FragColor = v_color;
+			}
+		}
+		""" @ glsl_fragment_shader
+
+	# Vertices coordinates
+	var coord = attributes["coord"].as(AttributeVec4) is lazy
+
+	# Should this program use the texture `texture`?
+	var use_texture = uniforms["use_texture"].as(UniformBool) is lazy
+
+	# Visible texture unit
+	var texture = uniforms["texture"].as(UniformSampler2D) is lazy
+
+	# Coordinates on the textures, per vertex
+	var tex_coord = attributes["tex_coord"].as(AttributeVec2) is lazy
+
+	# Color tint per vertex
+	var color = attributes["color"].as(AttributeVec4) is lazy
+
+	# Translation applied to each vertex
+	var translation = attributes["translation"].as(AttributeVec4) is lazy
+
+	# Rotation matrix
+	var rotation = uniforms["rotation"].as(UniformMat4) is lazy
+
+	# Scaling per vertex
+	var scale = attributes["scale"].as(AttributeFloat) is lazy
+
+	# Model view projection matrix
+	var mvp = uniforms["mvp"].as(UniformMat4) is lazy
+end
+
+redef class Point3d[N]
+	fun offset(x, y, z: Numeric): Point3d[Float]
```

Comment:
Because it's Float specific I prefer to leave it here. But we can move and update it when needed.